libvast/vast/plugin.hpp

```diff
@@ -0,0 +1,177 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/command.hpp"
+#include "vast/fwd.hpp"
+
+#include <caf/error.hpp>
+#include <caf/stream.hpp>
+#include <caf/typed_actor.hpp>
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+namespace vast {
+
+// -- plugin singleton ---------------------------------------------------------
+
+namespace plugins {
+
+/// Retrieves the system-wide plugin singleton.
+std::vector<plugin_ptr>& get() noexcept;
+
+} // namespace plugins
+
+// -- plugin version -----------------------------------------------------------
+
+/// The version of a plugin in format major.minor.patch.tweak.
+extern "C" struct plugin_version {
+  uint16_t major;
+  uint16_t minor;
+  uint16_t patch;
+  uint16_t tweak;
+};
+
+/// Checks if a version meets the plugin version requirements.
+/// @param version The version to compare against the requirements.
+bool has_required_version(const plugin_version& version) noexcept;
```

Comment:
Although in the medium- to long-term we certainly need a separate plugin versioning scheme (as soon as we have binary third-party plugins), for now we could probably save ourselves a lot of headaches and lingering compatibility bugs by requiring that all plugins must be built from the same commit as vast itself.

Reply:
We're keeping this as is, doesn't hurt to have this now already. We can always make a breaking change while it's labeled experimental.

Reply:
I was more thinking about the scenario where we unintentionally break ABI but forget to bump the major version of the plugin API, leading to strange memory corruption.

Actually, while writing this I noticed that the implementation of this function also seems off, at the least the major version should match exactly as opposed to being `<=`.